### PR TITLE
test(bc): fix wall-clock flakes via HUG_FAKE_CLOCK (UTC-fixed)

### LINKED ISSUE #200: test(unit): 'hug bc --point-to: auto-generated name is unique per minute' is a time-based flake
## What

The unit test `hug bc --point-to: auto-generated name is unique per minute` (`tests/unit/test_bc.bats:334`) is a **timing-dependent flake** that intermittently fails in CI.

## Evidence

CI run on [elifarley/hug-scm#199](https://github.com/elifarley/hug-scm/pull/199) (commit `c5ad53c`, run [29040406553](https://github.com/elifarley/hug-scm/actions/runs/29040406553)) failed with exactly this output:

```
not ok 1330 hug bc --point-to: auto-generated name is unique per minute in 311ms
# `assert_output --partial "Generated name existed; using"` failed
# -- output does not contain substring --
# substring:   Generated name existed; using
# output:
#   ℹ️ Info: Auto-generated branch name: v1.0.0.branch.20260709-1830
#   Switched to a new branch 'v1.0.0.branch.20260709-1830'
```

Note the generated name is `v1.0.0.branch.20260709-1830` — minute-granularity. The test ran near a minute boundary (18:30), and the two `hug bc --point-to` calls landed on different sides of the boundary.

## Root cause

The test creates branch `v1.0.0.branch.YYYYMMDD-HHMM` twice and asserts that the **second** invocation detects a name collision and prints `Generated name existed; using <suffix>` (line 350):

```bash
run hug bc --point-to v1.0.0
# ...
run hug bc --point-to v1.0.0
assert_output --partial "Generated name existed; using"
```

The collision detection only fires when both calls land in the **same minute**. If the minute rolls over between the two `hug bc` calls (probability scales with how close to `:SS=59` the test starts), the second call generates a fresh, non-colliding name and the assertion fails. This is a classic wall-clock-based test flake.

## Proposed fix (options, ranked)

1. **Inject a fixed clock** (preferred): make `git-bc` accept an env var like `HUG_FAKE_CLOCK=YYYYMMDD-HHMM` that the test sets to a deterministic value for both calls. Both calls then provably collide regardless of real wall time. Also fixes any sibling test in the "auto-generated name" family.
2. **Force a collision deterministically**: pre-create the first branch with the exact expected name before invoking `hug bc --point-to`, so the collision path always triggers.
3. **Loosen the assertion**: accept either the collision message OR a distinct second branch name — weakens the test and hides regressions in the collision-suffix logic, so the worst option.

## Reproduction

Non-deterministic; easiest to repro by running the test in a tight loop near `HH:MM:59`:

```bash
for i in $(seq 1 50); do
  make test-unit TEST_FILE=test_bc.bats TEST_FILTER="auto-generated name is unique per minute" || break
done
```

## Impact

- Blocks otherwise-green PRs (the docs-only [elifarley/hug-scm#199](https://github.com/elifarley/hug-scm/pull/199) was blocked by this single flake on its first CI run).
- Erodes trust in CI signal — devs learn to "just re-run" and may miss real failures.

## Tracking

Discovered while running `/pr-fix` on [elifarley/hug-scm#199](https://github.com/elifarley/hug-scm/pull/199). That PR is docs-only and unrelated to this flake.

### PR DESCRIPTION
**Pull request by elifarley**
## Summary
- Adds `git-config/lib/hug-clock` exposing `hug_clock_now` / `hug_clock_epoch` — UTC-fixed (`date -u`), GNU/BSD-portable, fail-safe, arg-validated.
- Registers `hug-clock` in `hug-common` with a TDD guard in `test_hug-common.bats`.
- Switches `git-bc`'s three direct `date` calls (name generation + collision suffix) to `hug_clock_now`.
- Pins BOTH flaky `test_bc.bats` tests (:334 the issue names, :136 Codex review surfaced) to `HUG_FAKE_CLOCK=946684800`; rewrites :334 with exact-UTC-name assertions and renames it.
- Zero behavior change when `HUG_FAKE_CLOCK` is unset (apart from the suffix timezone shifting local → UTC — documented as acceptable for a uniqueness hint).
- Hardened by adversarial Codex review: all 9 findings incorporated (TZ determinism, BSD portability, numeric-but-unformattable fallback, separate-stderr test, arg validation, exact-name assertions, registration TDD guard, fixed verification greps, second-flake discovery).

## Test plan
- [x] `make test-lib TEST_FILE=test_hug_clock.bats` (9 new lib tests, TZ-independent)
- [x] `make test-unit TEST_FILE=test_bc.bats` (full bc suite, incl. both pinned tests)
- [x] 50× loop on BOTH pinned tests passes 50/50 (the flake-repro gate from the issue)
- [x] `make sanitize-check` green (bash lint, python lint, typecheck)
- [ ] `make test` (full suite) green — note: one pre-existing env-related failure (`test_commit.bats:257` "hug c: surfaces commit hook failures") unrelated to this PR; the commit-msg hook doesn't fire under the local hug/git setup. Confirmed identical on origin/main.

Closes elifarley/hug-scm#200.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by elifarley** [comment]
Heads-up for reviewers: the one `make test` failure I called out in the PR body is a **pre-existing** environment-specific issue, now filed as elifarley/hug-scm#203.

**TL;DR:** `test_commit.bats:257` ("hug c: surfaces commit hook failures") fails on dev machines where `core.hooksPath` is set globally — the fresh test repo inherits the global hooksPath, so the `.git/hooks/commit-msg` it installs never fires. Passes on CI (no global hooksPath). Confirmed identical on `origin/main`, unrelated to this PR's changes.

The #202 work itself is green on all its own test files (`test_hug_clock.bats` 9/9, `test_hug-common.bats` 22/22, `test_bc.bats` 26/26, 50× flake loop on both pinned collision tests).

**Comment by elifarley** [comment]
# Code Roast: Ruthless Engineering Review

> Generated by the `code-roast` skill (Deep Hunt Protocol, P0–P6). Posted verbatim by request. Line numbers below are inflated vs. the actual 92-line file — see framing note at top for the corrected mapping.

---

## ANALYSIS SCOPE

**PR Title**: test(bc): fix wall-clock flakes via HUG_FAKE_CLOCK (UTC-fixed)
**PR Number**: #202
**Branch**: fix-bc-fake-clock-200
**Base Branch**: main
**Files Changed**: 9 files, +1393/-17 lines

**What's being reviewed**: Implementation of a `HUG_FAKE_CLOCK` environment variable override system to fix non-deterministic test flakes in `git-bc` (branch creation) command. The solution introduces:
- New `git-config/lib/hug-clock` library with UTC-fixed clock functions
- Registration of `hug-clock` in `hug-common` bootstrap
- Modification of `git-bc` to use the new clock functions instead of direct `date` calls
- Test updates to pin wall clock in flaky tests

**Key Implementation Points**:
- 91-line new library module
- 3 `date` call sites replaced in `git-bc`
- 9 new tests for the clock library
- 2 existing tests modified to use `HUG_FAKE_CLOCK`

**Entry Points Analyzed**:
- `hug_clock_now(fmt)` - formatted UTC time with override
- `hug_clock_epoch()` - epoch seconds with override
- `_hug_clock_format_epoch(epoch, fmt)` - platform-specific formatter
- `git-bc` name-generation block (lines 123-176)

**Domain Flags**: Time/clocks (primary), Error handling, Docs/claims

---

## CRITICAL ISSUES (Must Fix)

### 1. CRITICAL: Contract violation in `hug_clock_epoch` - bypasses fail-safe promise

**Location**: `git-config/lib/hug-clock` — `hug_clock_epoch()` function

**What**: The `hug_clock_epoch` function violates the documented fail-safe contract by returning invalid numeric epoch values verbatim without validation or warning, while the docstring promises "warns on stderr and falls back to the real UTC clock" for unformattable values.

**Why This Matters**: This is a silent contract violation that will cause production incidents. A user who sets `HUG_FAKE_CLOCK=99999999999999999999` (overflow value) expects the library to fall back to real time, but instead gets an invalid epoch that will break any system that consumes it (branch names, log timestamps, database records). The failure mode is silent - no warning, no error, just wrong data.

**Specific Problems**:
```bash
# From the docstring:
# Fail-safe:
#   If HUG_FAKE_CLOCK is set but invalid (non-numeric, or numeric-but-unformattable),
#   warns on stderr and falls back to the real UTC clock. Never crashes a calling
#   command over a bad env value.

# But the actual code in hug_clock_epoch():
hug_clock_epoch() {
  local fake="${HUG_FAKE_CLOCK:-}"
  if [[ -n "$fake" ]] && [[ "$fake" =~ ^[0-9]+$ ]]; then
    printf '%s\n' "$fake"    # ← NO VALIDATION, NO WARNING, NO FALLBACK
    return 0
  fi
  date -u +%s
}
```

Compare to `hug_clock_now()` which DOES implement the fail-safe contract correctly:
```bash
if [[ "$fake" =~ ^[0-9]+$ ]]; then
  local rendered
  if rendered=$(_hug_clock_format_epoch "$fake" "$fmt") && [[ -n "$rendered" ]]; then
    printf '%s\n' "$rendered"
    return 0
  fi
  # ← Warns and falls back if formatting fails
  printf 'hug-clock: HUG_FAKE_CLOCK=%q could not be formatted; using real clock\n' \
    "$fake" >&2
fi
```

**Fix**: `hug_clock_epoch` must match `hug_clock_now`'s fail-safe behavior:

```bash
hug_clock_epoch() {
  local fake="${HUG_FAKE_CLOCK:-}"
  if [[ -n "$fake" ]]; then
    if [[ "$fake" =~ ^[0-9]+$ ]]; then
      # Validate by attempting to format it (same as hug_clock_now)
      local rendered
      if rendered=$(_hug_clock_format_epoch "$fake" "%Y%m%d-%H%M") && [[ -n "$rendered" ]]; then
        printf '%s\n' "$fake"
        return 0
      fi
      printf 'hug-clock: HUG_FAKE_CLOCK=%q could not be formatted; using real clock\n' \
        "$fake" >&2
    else
      printf 'hug-clock: ignoring invalid HUG_FAKE_CLOCK=%q (expected unix epoch)\n' \
        "$fake" >&2
    fi
  fi
  date -u +%s
}
```

**Verification Trigger**:
```bash
# Set an overflow value that date cannot format
export HUG_FAKE_CLOCK=99999999999999999999
source git-config/lib/hug-clock
hug_clock_epoch  # Returns: 99999999999999999999 (WRONG - no warning)
```

**Expected**: Should warn "could not be formatted" and return current epoch.

**Blast Radius**: Every future caller of `hug_clock_epoch` (documented as ready for `git-tc`, `git-w-wip` adoption) will receive invalid timestamps when users set overflow epoch values. Silent data corruption in branch names, logs, or any time-stamped artifacts.

---

## MAJOR CONCERNS (Should Fix)

### 1. MAJOR: Inconsistent API - two functions, two different error handling strategies

**Location**: `git-config/lib/hug-clock` — both public functions

**What**: The library exports two functions (`hug_clock_now` and `hug_clock_epoch`) with identical purpose (honor `HUG_FAKE_CLOCK`) but completely different error handling behaviors. One is fail-safe, the other is not.

**Why This Matters**: API inconsistency is a maintenance nightmare. Future developers (including yourself 6 months from now) will reasonably assume both functions behave the same way. The pattern "set HUG_FAKE_CLOCK, get deterministic time" should work uniformly. Currently:
- `hug_clock_now`: Validates → Formats → Warns on failure → Falls back
- `hug_clock_epoch`: Validates regex → Returns verbatim (no format check)

This asymmetry will lead to bugs when code is refactored between the two functions.

**Specific Problems**:
```bash
# These SHOULD behave identically for the same input:
export HUG_FAKE_CLOCK=99999999999999999999

hug_clock_now "%Y%m%d-%H%M"
# Returns: (real UTC time) + stderr warning → FAIL-SAFE ✓

hug_clock_epoch
# Returns: 99999999999999999 → SILENTLY WRONG ✗
```

**Fix**: Apply the same fail-safe pattern to both functions. The CRITICAL issue fix above resolves this.

### 2. MAJOR: Missing test coverage for the contract violation

**Location**: `tests/lib/test_hug_clock.bats` — the "returns override verbatim" test

**What**: The test suite explicitly verifies the WRONG behavior for `hug_clock_epoch`. The test "returns override verbatim" confirms that invalid values are returned without validation, which contradicts the documented fail-safe contract.

**Why This Matters**: Tests codify behavior. When a test asserts "function returns garbage verbatim", it locks in a bug. Future developers will see this passing test and assume the behavior is intentional. The test should assert fail-safe behavior, not silent data corruption.

**Specific Problems**:
```bash
@test "hug_clock_epoch: returns override verbatim" {
  HUG_FAKE_CLOCK=946684800 run hug_clock_epoch
  assert_success
  assert_output "946684800"  # ← This is fine for VALID epochs
}
# But there's NO test for:
# HUG_FAKE_CLOCK=99999999999999999999  # INVALID epoch
# Expected: warning + real epoch
# Actual: returns invalid epoch (BUG)
```

**Fix**: Add a test case matching the `hug_clock_now` pattern:
```bash
@test "hug_clock_epoch: numeric-but-unformattable override falls back with warning" {
  HUG_FAKE_CLOCK=99999999999999999999 run --separate-stderr hug_clock_epoch
  assert_success
  [[ "$output" =~ ^[0-9]+$ ]]  # Should be a valid current epoch, not the overflow
  [[ "$stderr" == *"could not be formatted"* ]]
}
```

### 3. MAJOR: Docstring promises behavior not implemented

**Location**: `git-config/lib/hug-clock` — the fail-safe section of the docstring

**What**: The function docstring explicitly promises "If HUG_FAKE_CLOCK is set but invalid (non-numeric, or numeric-but-unformattable), warns on stderr and falls back to the real UTC clock." This promise is implemented in `hug_clock_now` but NOT in `hug_clock_epoch`.

**Why This Matters**: Docstrings are contracts. When code violates its own documentation, it creates a trap for every developer who reads the doc and trusts the API. The "why this matters" section of the CRITICAL issue applies here too - silent contract violations cause production incidents.

**Specific Problems**:
The docstring applies to the library as a whole, yet only one function implements it.

**Fix**: Either (a) implement the contract in `hug_clock_epoch` (recommended), or (b) explicitly document that `hug_clock_epoch` has different error handling. Option (a) is better for API consistency.

---

## MINOR OBSERVATIONS (Nice to Have)

### 1. MINOR: Platform detection could fail for non-standard date implementations

**Location**: `git-config/lib/hug-clock` — `_hug_clock_format_epoch()`

**What**: The platform detection tries GNU format first, then BSD format. If a system has neither (e.g., BusyBox, embedded systems), both attempts fail silently.

**Why This Matters**: Currently theoretical since the repo targets GNU/BSD systems per ADR-001. However, the silent failure (`2>/dev/null`) means diagnostic information is lost if the platform detection ever fails in the wild.

**Fix**: Consider logging a warning if BOTH formatters fail (though this may be over-engineering given the target platforms).

### 2. MINOR: Test comment contradicts test behavior

**Location**: `tests/lib/test_hug_clock.bats` — "returns override verbatim" test

**What**: Test comment says "returns override verbatim" but this is only correct for VALID epochs. For invalid epochs, the test should assert fallback behavior (once the CRITICAL issue is fixed).

**Fix**: Update comment to "returns valid override verbatim; falls back for unformattable values" after adding the missing test case.

---

## EDGE CASES ANALYSIS

### What Happens When:

1. **User sets `HUG_FAKE_CLOCK=garbage` (non-numeric)**:
   - `hug_clock_now`: Warns "ignoring invalid HUG_FAKE_CLOCK", returns real UTC ✓
   - `hug_clock_epoch`: Falls back to real epoch (regex rejects non-numeric) ✓
   - **Status**: CORRECT

2. **User sets `HUG_FAKE_CLOCK=99999999999999999999` (overflow)**:
   - `hug_clock_now`: Warns "could not be formatted", returns real UTC ✓
   - `hug_clock_epoch`: Returns overflow value verbatim ✗ **CRITICAL BUG**
   - **Status**: FAILS in `hug_clock_epoch`

3. **User sets `HUG_FAKE_CLOCK=-1` (negative)**:
   - `hug_clock_now`: Regex rejects, falls back to real UTC ✓
   - `hug_clock_epoch`: Regex rejects, falls back to real epoch ✓
   - **Status**: CORRECT

4. **User sets `HUG_FAKE_CLOCK=0` (epoch zero)**:
   - Both functions: Valid numeric, passes through ✓
   - **Status**: CORRECT

5. **User sets `TZ=Arctic/Longyearbyen` and `HUG_FAKE_CLOCK=946684800`**:
   - Both functions: Use `date -u`, ignore TZ ✓
   - **Status**: CORRECT (TZ independence verified by test)

6. **`date` command not installed on system**:
   - Both functions: Command fails, returns non-zero exit code
   - `git-bc` behavior: Would crash under `set -euo pipefail`
   - **Status**: EXISTING CONSTRAINT (date is required)

7. **User sets empty `HUG_FAKE_CLOCK=`**:
   - Both functions: Treated as unset (fallback to real time) ✓
   - **Status**: CORRECT

8. **Concurrent calls from multiple `git-bc` processes**:
   - Each reads `HUG_FAKE_CLOCK` independently (no shared state)
   - If set, all processes use the same frozen instant ✓
   - **Status**: CORRECT (no concurrency issues)

9. **User calls `hug_clock_now` with empty format string**:
   - Arg validation: `if [[ $# -ne 1 || -z "${1:-}" ]]` → returns usage error ✓
   - **Status**: CORRECT (arg validation works)

10. **Mid-batch abort during collision loop in `git-bc`**:
    - Loop uses `git show-ref` for existence check (no side effects)
    - Abort is safe - no partial state to clean up ✓
    - **Status**: CORRECT

---

## SUMMARY

### The Good

- **Testability achievement**: The `HUG_FAKE_CLOCK` override successfully makes wall-clock-dependent tests deterministic. The 50× loop verification in the plan is excellent.
- **TZ independence**: Correctly uses `date -u` throughout, verified by explicit TZ test (`TZ=America/Bahia` test passes).
- **Platform portability**: GNU/BSD date detection is well-implemented and tested.
- **Comprehensive test coverage**: 9 test cases for the clock library cover normal operation, TZ independence, invalid inputs, and overflow scenarios (for `hug_clock_now`).
- **TDD registration guard**: The `test_hug-common.bats` test ensuring `hug-clock` is loaded is exactly the right pattern.
- **Clean git-bc integration**: The 3 call-site replacements are minimal and well-commented.
- **Fail-safe design for `hug_clock_now`**: The warning-on-stderr + fallback behavior is production-grade.

### The Bad

- **CRITICAL contract violation**: `hug_clock_epoch` bypasses the fail-safe promise documented in the library's own docstring. Invalid numeric epochs are returned verbatim without warning or fallback.
- **MAJOR API inconsistency**: Two functions with the same purpose have wildly different error handling strategies.
- **MAJOR missing test coverage**: The test suite locks in the buggy behavior instead of asserting the documented fail-safe contract.
- **Silent data corruption risk**: Future adopters of `hug_clock_epoch` (documented for `git-tc`, `git-w-wip`) will receive invalid timestamps when users set overflow epoch values.

### Priority

1. **Fix the CRITICAL contract violation in `hug_clock_epoch`** - this is a production bug waiting to happen
2. **Add the missing test case for overflow epochs** - prevents regression
3. **Align both functions' error handling** - API consistency is critical for maintainability

### Simplification Opportunities

If this had to be maintained for one year, I would simplify:

1. **Single function instead of two**: `hug_clock_now "%s"` can return epochs, eliminating `hug_clock_epoch` entirely. The epoch-only function adds surface area without solving a distinct problem.

2. **Centralize validation**: Both functions duplicate the `[[ "$fake" =~ ^[0-9]+$ ]]` regex check. Extract to `_hug_clock_validate_and_format` helper that returns both the formatted value AND a validation flag.

3. **Simplify the formatter**: The `_hug_clock_format_epoch` function could be inlined into `hug_clock_now` since it's only called from one place. The abstraction doesn't pull its weight.

---

**Reviewer**: Code Roast Skill (Deep Hunt Protocol)
**PR**: elifarley/hug-scm#202
**Analysis Depth**: Full P0-P6 passes completed, Verification Gate applied to all findings

File: tests/lib/test_hug_clock.bats [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  assert_output --partial "usage"
+}
+
+@test "hug_clock_now: returns real UTC time when override unset" {
+  unset HUG_FAKE_CLOCK
+  run hug_clock_now "%Y%m%d-%H%M"
+  assert_success
+  # Must match YYYYMMDD-HHMM and equal date -u (UTC), not local time.
+  [[ "$output" =~ ^[0-9]{8}-[0-9]{4}$ ]]
+  assert_equal "$output" "$(date -u +"%Y%m%d-%H%M")"
```
[the remark this answers is not in the document]
Comment (elifarley) [reply tests/lib/test_hug_clock.bats]: Fixed — removed the exact-equality assert_equal with the second date call. The regex on the preceding line already proves the YYYYMMDD-HHMM format, and the TZ-independence test (line 26) covers the UTC contract. This eliminates the minute-boundary straddle risk.

**Comment by elifarley** [comment]
## Code Roast Response

All findings evaluated and addressed. Triage summary:

### CRITICAL + MAJOR (contract violation in `hug_clock_epoch`)
**Implemented.** The roast correctly identified that `hug_clock_epoch` lacked the format-check validation that `hug_clock_now` already had. Both functions now share the same fail-safe pattern: attempt `_hug_clock_format_epoch`, warn on stderr, fall back to real clock.

One pushback on severity framing: `HUG_FAKE_CLOCK` is a **test-only mechanism** set by test scripts, never by end users. The "production incident" scenario (a user setting `HUG_FAKE_CLOCK=99999999999999999999` in production) is not a realistic threat model. That said, the **API inconsistency was real** — two functions with the same purpose should behave identically — so the fix was warranted regardless.

Changes:
- `hug_clock_epoch` now validates through `_hug_clock_format_epoch` before accepting the override (same gate as `hug_clock_now`).
- Added Fail-safe docstring to `hug_clock_epoch` matching `hug_clock_now`'s contract.
- Added test: `hug_clock_epoch: numeric-but-unformattable override falls back with warning`.
- Upgraded `hug_clock_epoch: invalid override` test to assert stderr warning via `--separate-stderr`.

### Codex P2 (wall-clock comparison in test)
**Implemented.** Removed the `assert_equal` with the second `date -u` call on test_hug_clock.bats:20. The regex `^[0-9]{8}-[0-9]{4}$` already proves format correctness, and the TZ-independence test covers the UTC contract. This eliminates the minute-boundary straddle risk the Codex review identified.

### MINOR (platform detection silent failure)
**Not addressed.** Agree it's theoretical (per ADR-001, the repo targets GNU/BSD only). Adding a warning for a scenario that cannot occur on supported platforms would be over-engineering.

### Simplification suggestions (single function, centralize validation)
**Not adopted.** `hug_clock_now` and `hug_clock_epoch` serve distinct callers: `hug_clock_now` returns formatted strings (used in branch names), `hug_clock_epoch` returns raw integers (future use in timestamps). Conflating them via `hug_clock_now "%s"` would hide the semantic distinction and force every caller to know the magic format string. The current two-function API is clearer at call sites.

All 10 library tests pass, all 26 bc unit tests pass, `make sanitize` clean. Commit: c4b2f9c.
